bpf/bpf_lxc.c

```diff
@@ -1768,7 +1779,16 @@ ipv4_policy(struct __ctx_buff *ctx, int ifindex, __u32 src_label, enum ct_status
 #endif /* ENABLE_PER_PACKET_LB && !DISABLE_LOOPBACK_LB */
 
 #ifdef ENABLE_NODEPORT
-	if (ret == CT_NEW || ret == CT_REOPENED) {
+	if (tuple->nexthdr == IPPROTO_TCP && has_l4_header) {
+		if (ctx_load_bytes(ctx, l4_off + 12, &tcp_flags, 2) < 0)
+			return DROP_CT_INVALID_HDR;
+	}
+	if (ret == CT_NEW || ret == CT_REOPENED ||
+	    /* In case a DSR flow hits a stale nodeport ct entry, override it
+	     * so that a reply packet will be SNATed properly
+	     */
+	    (ret == CT_ESTABLISHED && ((tuple->nexthdr != IPPROTO_TCP || !has_l4_header) ||
+	    ((tcp_flags.value & TCP_FLAG_SYN) && !(tcp_flags.value & TCP_FLAG_ACK))))) {
```

Comment:
I think the overriding of the stale CT entries should be relevant only for the DSR case. In the case of SNAT, it's enought to set `node_port` in the CT entry which should be set in the previous CT entry. Therefore, this `SYN && !ACK` check should be protected with `#ifdef ENABLE_DSR`.

Reply:
@ysksuzuki Could you elaborate on the logic when a packet hits this flow under !has_l4_header?

Reply:
@borkmann This is a mistake. I think we should override the stale CT entries when it's SYN, not a fragmented IP packet without an l4 header. I will fix it.

Reply:
@borkmann This is a mistake. I think we should not override the stale CT entries when it's a fragmented IP packet without an l4 header. I will fix it.